### ISSUE TITLE
Why does the implementation here reduce memory?

### ISSUE THREAD

**Issue by lartpang** [microsoft/vision-longformer#7]
@jwyang @pzzhang 

Thanks for your great work!

> In the sliding chunk approach, to achieve a conv-like local attention mechanism with window size $2 w + 1$, we split the feature map into chunks with size $w \times w$. Each chunk only attends to itself and its 8 neighbor chunks. The Pytorch Autograd will save 9 copies of the feature map (9 nodes in the computing graph) for automatic back-propagation, which is not time/memory efficient. The SCw/Handgrad version defines a customized torch.autograd.Function with hand-written backward function, which greatly saves the memory usage and also speeds up the algorithm, as shown in figures above. We would like to point out that the memory usage of the SCw/Handgrad version is nearly optimal (very close to that of the cuda_kernel).

In my experiment, the memory is really much lower.
But I am very interested in the principle behind it.

> The Pytorch Autograd will save 9 copies of the feature map (9 nodes in the computing graph) for automatic back-propagation...

How to understand this? 

> The SCw/Handgrad version defines a customized torch.autograd.Function with hand-written backward function, which greatly saves the memory usage and also speeds up the algorithm

How does the implementation here avoid this? 
I am very curious about this and hope to get a more detailed explanation. 
Thank you very much!

**Comment by pzzhang** [comment]
"The Pytorch Autograd will save 9 copies of the feature map (9 nodes in the computing graph) for automatic back-propagation..."
If we use Pytorch Autograd, it will try to save too many intermediate variables for backpropagation. 

If we use SCw/Handgrad version, we save much less intermediate variables for backpropagation.

**Comment by lartpang** [comment]
@pzzhang Thanks for your reply!

Can the results of these intermediate variables be simply interpreted as the tensors stored in the `ctx.save_for_backward()` of the `Function` corresponding to the internal operation?

When we manually construct the backpropagation process, we can directly obtain the gradient of each leaf node from the operational relationship at the level of the entire module.
Therefore, it is no longer necessary to save the results of fine-grained operations, such as the output of the internal convolutional layer.
However, PyTorch's own autograd is based on more fine-grained operations, which actually saves more intermediate activation values.

**Comment by pzzhang** [comment]
@lartpang  yeah, your understanding is correct. For the vision longformer case, the torch.roll operation is one of the fine-grained operations that saves intermediate activation values but can be avoided with the SCw/Handgrad version.

**Comment by lartpang** [comment]
@pzzhang Thanks!
